remixer_core/src/main/java/com/google/android/libraries/remixer/Remixer.java

```diff
@@ -77,47 +99,28 @@ public static Remixer getInstance() {
   public void addItem(RemixerItem remixerItem) {
     List<RemixerItem> listForKey = getItemsWithKey(remixerItem.getKey());
     List<RemixerItem> itemsToRemove = new ArrayList<>();
+    Object parentObject = remixerItem.getParentObject();
     for (RemixerItem existingItem : listForKey) {
-      existingItem.assertIsCompatibleWith(remixerItem);
-      if (!existingItem.hasParentObject()) {
-        // The parent activity has been reclaimed by the OS already. It has no callback and it's
-        // still around for keeping the value in sync, saving and checking consistency across types.
-        // Since we're adding a new item that has been asserted to be compatible, it is not
-        // necessary to keep this instance around.
-        itemsToRemove.add(existingItem);
-      } else {
-        // The parent activity is still alive and kicking.
-        if (existingItem.isParentObject(remixerItem.getParentObject())) {
-          // An object with the same key for the same parent object, this shouldn't happen so throw
-          // an exception.
-          throw new DuplicateKeyException(
-              String.format(
-                  Locale.getDefault(),
-                  "Duplicate key %s being used in class %s",
-                  remixerItem.getKey(),
-                  remixerItem.getParentObject().getClass().getCanonicalName()
-              ));
-        }
+      if (parentObject != null && parentObject == existingItem.getParentObject()) {
+        // An object with the same key for the same parent object, this shouldn't happen so throw
+        // an exception.
+        throw new DuplicateKeyException(
+            String.format(
+                Locale.getDefault(),
+                "Duplicate key %s being used in class %s",
+                remixerItem.getKey(),
+                remixerItem.getParentObject().getClass().getCanonicalName()
+            ));
```

Comment:
could this logic also belong in synchronizationMechanism? Could listForKey belong in synchronizationMechanism?

synchronizationMechanism is supposed to "Takes over the responsibility of making sure that RemixerItems being added to remixer are compatible with what is already there from the Remixer class."